libraries/AP_Scheduler/AP_Scheduler.cpp

```diff
@@ -240,6 +241,23 @@ void AP_Scheduler::run(uint32_t time_available)
         _spare_ticks /= 2;
         _spare_micros /= 2;
     }
+
+    // soak up any spare loop time with provided function
+    if (_looptime_soak_fn != nullptr) {
+        uint32_t soaked_time_ms = 0;
+        while (time_available > 1000 && soaked_time_ms < 5000) {
+            // get remaining time available for this loop
+            _task_time_started = AP_HAL::micros();
+            _task_time_allowed = time_available; // FIXME: add margin here?
+            _looptime_soak_fn();
+            const uint32_t time_taken = AP_HAL::micros() - _task_time_started;
+            soaked_time_ms += time_taken;
+            if (time_taken < 1000) {
+                hal.scheduler->delay_microseconds(1000-time_taken);
```

Comment:
At a minimum we need to be releasing and regaining the scheduler semaphore here, alternatively we should always release it and sleep for just a moment so we don't starve a low priority thread who is attempting to get access to it.